### ISSUE TITLE
Daly BMS via can bus

### ISSUE THREAD

**Issue by SamuelBrucksch** [Louisvdw/dbus-serialbattery#167]
Hi,

Some of the Daly BMS support a connection via can bus. That should be considerably faster than the serial connection. The protocol is also open.

So i was wondering if - technically - this could be implemented into this library as well. I could work on it and try it myself as I have such a Daly BMS with can bus, i just want to know if there could be any problems I did not see yet.

Each BMS has its own implementation of the communication, so a new BMS type Daly_CAN could be added that gets the data via can bus, however some logic needs to detect/search for the can bus too, right?

As the cerbo for example already has VE.Bus hardware, do you by chance know, if that could be used directly? Then a RJ45 to Daly cable would be enough and I could start working on this...

**Comment by SamuelBrucksch** [comment]
Hm looks like CAN on daly is a gamble as noone really knows if they have can or not. Read from some guys, that only RS485 is available but not can bus, although it is printed on the case and also marked on the cables...

**Comment by geromd** [comment]
@SamuelBrucksch  Yes, some Daly BMS also have CAN-Bus. Before I decided for the RS485 way I already connected the CAN Port with RJ45 to the Vicron BMS CAN. Due that communication packages were counted, I know that the CAN-Connection was working, but the Victron did not recognizied the Daly BMS on CAN dur to missing driver.

**Comment by SamuelBrucksch** [comment]
How did you know can was working if you could not see anything except counting packages? 

As soon as you connect, Venus OS sends out messages on CAN that fit to the selected CAN profile. Those should be counted on TX.

Did OU also see messages on RX?

Venus OS has some CAN CLI tools like can send, candump that you can use to send messages and see the traffic. When I tried it, i could only see the outgoing messages but could not get any reply from the Daly even when I sent requests to it via cansend.

**Comment by SamuelBrucksch** [comment]
Ok can now connect to Daly BMS via canbus. Here a sample for cells voltage:
```
  candump can0
  can0  18950140   [0] 
  can0  18954001   [8]  01 0D 7D 0D 7E 0D 7E 58
  can0  18954001   [8]  02 0D 7E 0D 7F 0D 7E 58
  can0  18954001   [8]  03 0D 7E 0D 7F 0D 7E 58
  can0  18954001   [8]  04 0D 7E 0D 7E 0D 7E 58
  can0  18954001   [8]  05 0D 7E 0D 7F 0D 7F 58
  can0  18954001   [8]  06 0D 7F 0D 7F 0D 7F 58
```

Can needs to be set to 250000, requests look like this:
```
cansend can0 18950140#
```

So the request ID is 18 + Data ID + BMS ID (01) + Response ID (40)

**Comment by SamuelBrucksch** [comment]
Looking good, first try with python-can lib:
```
INFO:SerialBattery:Starting dbus-serialbattery
INFO:SerialBattery:dbus-serialbattery v0.12
INFO:SerialBattery:No Port needed
INFO:SerialBattery:Testing DalyCAN
18944001: bytearray(b'\x10\x01\x00\x00\x02\x00\x00\x00')
INFO:SerialBattery:DalyBMS 16 cells
INFO:SerialBattery:Connection established to DalyCAN
INFO:SerialBattery:DeviceInstance = 2
INFO:SerialBattery:com.victronenergy.battery.USB9
INFO:SerialBattery:Battery connected to dbus from /dev/tty/USB9
INFO:SerialBattery:DeviceInstance = 2
INFO:SerialBattery:Changed DeviceInstance = 2
INFO:SerialBattery:read_soc_data
18904001: bytearray(b'\x02\x12\x00\x00t\x9b\x03O')
INFO:SerialBattery:voltage: 530, current: -14.9, soc: 847
```

All data requested is from can bus

**Comment by SamuelBrucksch** [comment]
Oh yeah:
<img width="498" alt="image" src="https://user-images.githubusercontent.com/8776268/182229157-5583a728-f41a-477a-9856-3b3b8539263b.png">

It's fully working now. Needs some cleanup, but will create a PR soon, so we can check if everything is ok. Adoption of Daly to Daly CAN was actually quite easy as the protocol behind is still the same.

**Comment by geromd** [comment]
Wow, this would really be great if it can be added to the driver, to work with the CAN-Bus.

**Comment by SamuelBrucksch** [comment]
https://github.com/Louisvdw/dbus-serialbattery/pull/169

**Comment by SamuelBrucksch** [comment]
Cell voltages from last 6h:
<img width="694" alt="image" src="https://user-images.githubusercontent.com/8776268/182308713-e1d12fd6-e83d-47f2-a97b-2d94951bdfea.png">
Seems to be stable and also the values look very nice now. No weird values anymore.

**Comment by pau1phi11ips** [comment]
@SamuelBrucksch That looks good! What are you using to generate the cell voltages graph?

**Comment by SamuelBrucksch** [comment]
I use openhab which has Grafana integrated for visualization of the values that openhab gets. Requires influxdb additionally to that:
https://community.openhab.org/t/influxdb-grafana-persistence-and-graphing/13761

**Comment by pau1phi11ips** [comment]
Thanks, it gave me a head start. Found the Victron Grafana docker and it was super simple to install that and get the stats.

https://github.com/victronenergy/venus-docker-grafana

![image](https://user-images.githubusercontent.com/677264/187649124-cc77bf91-6463-46c4-93e0-461ffe326b57.png)

**Comment by RedCross0815** [comment]
Sorry for my stupididi but i need help. Iam reading and trying for hours now to get the daly bms via can to run. 

I get these meassage via can bus, as in the thread before:
![putty](https://user-images.githubusercontent.com/45964713/190728707-cda49f70-ee33-47a0-8c7a-93e0c78ef7d9.png)

But how do i get the bms into the device List like this?

![grafik](https://user-images.githubusercontent.com/45964713/190728589-20b5a021-792c-4e24-b43a-3240c839dce8.png)

I did all the installation shown in the video and the text, but i wont work...

**Comment by SamuelBrucksch** [comment]
You need two terminals to test if the CAN connection works. One for candump and one fro cansnd. Your cansend snippet does not show any Daly messages. Also you need to execute the dbus-serialbattery code manually, as there is no CAN auto detection yet.

**Comment by mr-manuel** [comment]
What is the current state of this?
